Implement RelMeAuth

capjamesg opened the pull request:
This PR contains an implementation of the RelMeAuth algorithm, commonly used in addition to IndieAuth to create a system wherein someone can authenticate with a service using a URL.

Appropriate documentation is also included in this PR.

I shall issue another commit at a later date that addresses the lint issue picked up during the CI stage, before this PR is merged into the main branch.

capjamesg committed:
implement relmeauth

capjamesg committed:
add documentation for relmeauth, check for links in link header, check for two way links

capjamesg committed:
add relmeauth test case

jamesvandyne commented on src/indieweb_utils/indieauth/relmeauth.py:
```diff
@@ -0,0 +1,66 @@
+
+    rel_me_links = mf2_data["rels"].get("me", [])
+
+    valid_rel_me_links = []
```
We could simplify slightly if we make valid_rel_me_links a `set` and just turn it into a list on return.

```suggestion
    valid_rel_me_links: Set[str] = set()
```

jamesvandyne commented on src/indieweb_utils/indieauth/relmeauth.py:
```diff
@@ -0,0 +1,66 @@
+            if item.get("href") == canonical_url:
+                canonical_link = canonicalize_url(link, link_domain)
+
+                valid_rel_me_links.append(canonical_link)
```
Assuming this is a set...
```suggestion
                valid_rel_me_links.add(canonical_link)
```

jamesvandyne commented on src/indieweb_utils/indieauth/relmeauth.py:
```diff
@@ -0,0 +1,66 @@
+    # remove any duplicate links
+    valid_rel_me_links = list(set(valid_rel_me_links))
+
+    return valid_rel_me_links
```
As valid_rel_me_links is already a set this can just be
```suggestion
    return list(valid_rel_me_links)
```

jamesvandyne commented on tests/test_indieauth/test_relmeauth.py:
```diff
@@ -0,0 +1,19 @@
+        return get_valid_relmeauth_links
+
+    def test_rel_me_auth_link_discovery(self, target):
+        # NOTE: This test will take a few seconds as it executes multiple requests.
```
We should pass in a hardcoded template with some hard-coded responses for each rel link and add those to responses objects so this test does not make use of the network. Similar as to how the [test_profile](https://github.com/capjamesg/indieweb-utils/blob/ca0ffe5f82e5fa24df61b9d5b58210f7bc9179ac/tests/test_indieauth/test_profile.py#L28) tests work.

jamesvandyne replied:
@capjamesg This should follow the other tests and use `responses` so it doesn't make a network call. 

```python
import responses

def test_rel_me_auth_link_discovery(self, target, index):
    url = "https://jamesg.blog"
    responses.add(responses.Response(method="GET", url=url, body=index))
    assert target(url) == "https://indieweb.social/@capjamesg"]
```
  
The `index` parameter in the function tells pytest to load the `index` fixture, which is located in [tests/conftest.py](https://github.com/capjamesg/indieweb-utils/blob/main/tests/conftest.py#L16), which simply reads `tests/fixtures/index.html` (which is a snapshot of the html on your index.html)

jamesvandyne commented on src/indieweb_utils/indieauth/relmeauth.py:
```diff
@@ -0,0 +1,66 @@
+        if link_valid.status_code != 200:
+            continue
+
+        parsed_page = BeautifulSoup(link_valid.text, "html.parser")
```
Good work on this function. Easy to understand and follow. Well done.

capjamesg replied:
Thank you!

jamesvandyne reviewed (approved):
LGTM!

capjamesg committed:
address linting issues, change rel me links to a set

capjamesg committed:
add documentation to relmeauth function, add require_rel_me_link_back property

capjamesg committed:
fix linting issues

capjamesg commented:
@jamesvandyne Mind giving this another quick review?

capjamesg committed:
Merge branch 'main' into relmeauth

jamesvandyne commented on src/indieweb_utils/indieauth/relmeauth.py:
```diff
@@ -0,0 +1,82 @@
+            if not item.get("rel") and require_rel_me_link_back is True:
+                continue
+
+            if "me" not in item.get("rel", "") and require_rel_me_link_back is True:
```
Should this be

```
if require_rel_me_link_back and item.get("rel", "").lower() != "me"
```

So it only matches against exact matches of me? We should probably normalize it as well so if someone has rel="Me" or such it still works.

jamesvandyne reviewed (approved):
LGTM w/ 1 should in the unit tests.

capjamesg committed:
use responses instead of http request for rel me auth test

capjamesg committed:
Merge branch 'relmeauth' of https://github.com/capjamesg/indieweb-utils into relmeauth

capjamesg commented:
I have addressed the feedback above and have used responses for our testing. I am going to merge this into main. Thank you for your assistance @jamesvandyne!
